### PR TITLE
Remove `MaxAndArgmax` `Op`

### LINKED ISSUE #765: Remove `MaxAndArgmax`
`MaxAndArgmax` is redundant and serving a very questionable role.

First, its implementation(s) perform distinct max and arg-max steps, so there doesn't appear to be any sort of efficiency gain behind its combined-`Op` approach.  

Second, there are already distinct `Max` and `ArgMax` `Op`s, but they're only used when an "uncanonicalization" rewrite replaces `MaxAndArgmax` with them.  This is unnecessarily roundabout, especially for a something of questionable implementation value.  If anything, this situation work **the opposite way**: i.e. replace `Max` and `ArgMax` nodes with a single `MaxAndArgmax`.

Anyway, it seems like all we need to do is copy over the `MaxAndArgmax.[R_op, grad]` implementations to their respective `Op`s.

### PR DESCRIPTION
**Pull request by aerubanov**
Closes #765

### CODE REVIEW

File: aesara/tensor/math.py [hunk 1]
```diff
@@ -467,6 +251,9 @@ def infer_shape(self, fgraph, node, shapes):
         )
         return [rval]
 
+    def R_op(self, inputs, eval_points):
+        raise ValueError("Argmax is not differentiable operation")
```
Comment (brandonwillard) [inline aesara/tensor/math.py:255]:
```suggestion
        raise ValueError("Argmax is not a differentiable operation")
```

File: aesara/tensor/math.py [hunk 2]
```diff
@@ -629,6 +412,54 @@ class Max(NonZeroCAReduce):
     def __init__(self, axis):
         super().__init__(aes.scalar_maximum, axis)
 
+    def R_op(self, inputs, eval_points):
+        if eval_points[0] is None:
+            return
+        if self.axis[0] > 1:
+            raise ValueError("R_op supported for max only when " " axis is 0 or 1")
```
Comment (brandonwillard) [inline aesara/tensor/math.py:419]:
```suggestion
            raise ValueError("R_op supported for max only when axis is 0 or 1")
```

File: aesara/tensor/math.py [hunk 3]
```diff
@@ -629,6 +412,54 @@ class Max(NonZeroCAReduce):
...
+        if inputs[0].ndim != 2:
+            raise ValueError("R_op supported for max only when " " input is a matrix")
```
Comment (brandonwillard) [inline aesara/tensor/math.py:421]:
```suggestion
            raise ValueError("R_op supported for max only when input is a matrix")
```

File: aesara/tensor/math.py [hunk 4]
```diff
@@ -629,6 +412,54 @@ class Max(NonZeroCAReduce):
...
+        if self.axis[0] == 0:
+            return eval_points[0][max_pos, arange(eval_points[0].shape[1])]
+        else:
+            return eval_points[0][arange(eval_points[0].shape[0]), max_pos]
+
+    def grad(self, inp, grads):
+        # The strict sense mathematical gradient of the maximum function is
+        # not calculated here for it is not defined at every point where some
+        # coordinates are identical. However, since the latter set has null
+        # Lebesgue measure, the result may be interpreted as weak gradient.
```
Comment (brandonwillard) [inline aesara/tensor/math.py:432]:
```suggestion
        # Lebesgue measure, the result may be interpreted as a weak gradient.
```

File: aesara/tensor/opt_uncanonicalize.py [hunk 5]
```diff
@@ -38,31 +38,31 @@
...
+#        axis = node.op.get_params(node)
+#        if len(fgraph.clients[node.outputs[1]]) == 0:
+#            new = Max(axis)(node.inputs[0])
+#            copy_stack_trace(node.outputs[0], new)
+#            return [new, None]
+#
+#        if len(fgraph.clients[node.outputs[0]]) == 0:
+#            new = Argmax(axis)(node.inputs[0])
+#            copy_stack_trace(node.outputs[0], new)
+#            return [None, new]
```
Comment (brandonwillard) [inline aesara/tensor/opt_uncanonicalize.py:65]:
```suggestion
```

File: tests/link/test_jax.py [hunk 6]
```diff
@@ -478,18 +476,18 @@ def assert_fn(x, y):
...
+#    reason="Omnistaging cannot be disabled",
+# )
+# def test_jax_basic_multiout_omni():
+#    # Test that a single output of a multi-output `Op` can be used as input to
+#    # another `Op`
+#    x = dvector()
+#    mx, amx = MaxAndArgmax([0])(x)
+#    out = mx * amx
+#    out_fg = FunctionGraph([x], [out])
+#    compare_jax_and_py(out_fg, [np.r_[1, 2]])
```
Comment (brandonwillard) [inline tests/link/test_jax.py:490]: Take a look at how the `MyMultiOut` `Op` is used in `tests.link.test_numba` for an example of what we could do to replace this multi-output `Op` test.

File: tests/link/test_numba.py [hunk 7]
```diff
@@ -2112,51 +2112,6 @@ def test_Softplus(x, exc):
...
-    cm = contextlib.suppress() if exc is None else pytest.warns(exc)
-    with cm:
-        compare_numba_and_py(
-            g_fg,
-            [
-                i.tag.test_value
-                for i in g_fg.inputs
-                if not isinstance(i, (SharedVariable, Constant))
-            ],
-        )
```
Comment (brandonwillard) [inline tests/link/test_numba.py:2157]: We can add a new `pytest.mark.parametrize` and parameter that cycles through the two now independent `Op`s and runs the same tests.

File: tests/tensor/nnet/test_basic.py [hunk 8]
```diff
@@ -906,71 +895,6 @@ def validate_grad_graph(func):
             assert softmax_grad_legacy not in ops
 
 
-def test_argmax_pushdown():
```
Comment (brandonwillard) [inline tests/tensor/nnet/test_basic.py:909]: It should be possible to refactor these so that they work with each `Op` separately.

File: tests/tensor/test_math.py [hunk 9]
```diff
@@ -745,400 +733,9 @@ def test_isnan():
         f([[0, 1, 2]])
 
 
-class TestMaxAndArgmax:
```
Comment (brandonwillard) [inline tests/tensor/test_math.py:748]: These tests need to be converted to work with the two `Op`s , as well.

File: tests/tensor/test_opt_uncanonicalize.py [hunk 10]
```diff
@@ -24,28 +22,6 @@
 from tests.link.test_link import make_function
 
 
-class TestMaxAndArgmax:
```
Comment (brandonwillard) [inline tests/tensor/test_opt_uncanonicalize.py:27]: Same here.

File: tests/tensor/test_math_opt.py [hunk 11]
```diff
@@ -4079,7 +4079,7 @@ def check_max_log_sum_exp(x, axis, dimshuffle_op=None):
 
         # in mode FAST_COMPILE, the optimisations don't replace the
         # MaxAndArgmax op.
-        if isinstance(node.op, MaxAndArgmax):
+        if isinstance(node.op, Max, Argmax):
```
Comment (brandonwillard) [inline tests/tensor/test_math_opt.py:4082]:
```suggestion
        if isinstance(node.op, (Max, Argmax)):
```
This appears to be the cause of the current CI error.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Looks like this is making good progress; much appreciated!

**Comment by aerubanov** [comment]
@brandonwillard,  thanks for the review! I will work on the requested changes.

**Comment by aerubanov** [comment]
Hi @brandonwillard! I have the test that fails because `max()` here https://github.com/aerubanov/aesara/blob/b310d4dc10c959e5b7917159dea8397aceaba56f/aesara/tensor/math.py#L561 does not work correctly when the input tensor has a value that is the maximum of the `uint64` type. I don't understand why this happens and what needs to be changed to fix it. Could you take a look when you have a chance?

**Review by brandonwillard — COMMENTED** [review]
I just made some small fixes and rebased onto `main`.  The errors I'm seeing now are due to the unfinished Numba implementations.  ~Is that older, more cryptic error still present?~ (Nevermind, I see it in `tests.tensor.test_math.TestMinMax.test_uint`.  I'll look into it.)

Since these commits need to be restructured/squashed anyway, it would be good to reorganize them so that the `Max` and `ArgMax` updates and their corresponding tests are added first (e.g. in a single commit or one for each `Op` along with their JAX and Numba implementations), then the removal/replacement of `MaxAndArgmax` in another commit.

**Comment by aerubanov** [comment]
@brandonwillard Thanks for the review! I'll be working on those changes.

**Comment by aerubanov** [comment]
Hi @brandonwillard ! I seem to be able to fix the errors that occurred after removing `MaxAndArgmax`. Now I'm going to edit the commit history. But I would also like to ask you to review my changes and pay attention to the following points:
- I implemented `Max` as `COp` using `MaxAndArgmax` code, because it allows me to fix `tests.tensor.test_math.TestMinMax.test_uint` (see previous comments). But may be it is not the best option;
- I keep  `max_and_argmax()` because it is part of API - now it create `Max` and `Argmax` nodes. Do we want to keep this function or I should make it deprecated?

File: tests/tensor/test_math_opt.py [hunk 12]
```diff
@@ -3596,8 +3596,8 @@ def test_local_reduce_broadcast_some_0(self):
             at_all,
             at_any,
             prod,
-            at_max,
-            at_min,
+            # at_max,
+            # at_min,
```
Comment (brandonwillard) [inline tests/tensor/test_math_opt.py:3600]: Why have the `max` and `min` functions been removed from `TestLocalReduce`?

File: aesara/tensor/math.py [hunk 13]
```diff
@@ -467,6 +467,9 @@ def infer_shape(self, fgraph, node, shapes):
         )
         return [rval]
 
+    def R_op(self, inputs, eval_points):
+        raise ValueError("Argmax is not a differentiable operation")
+
```
Comment (brandonwillard) [inline aesara/tensor/math.py:256]:
```suggestion
```
There's already a `grad` implementation&mdash;albeit effectively non-functional&mdash;so I don't know if it helps to have an `R_op` like this.  Also, I don't think a `ValueError` would be the best here.

If a gradient is undefined, I believe we should return an `aesara.gradient.grad_undefined`-generated `NullType`.  Same with unimplemented gradients: we should use `aesara.gradient.grad_not_implemented`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Looks like an extra file got added: `aesara/tensor/\`.

File: tests/tensor/test_math_opt.py (hunk 12 above)
Comment (aerubanov) [reply tests/tensor/test_math_opt.py:3600]: @brandonwillard I moved the `MaxAndArgmax` functionality to `Max` and now `Max` is implemented as a `COp`, so the tests for `CAReduce` will not work with `Max`.

**Comment by aerubanov** [comment]
@brandonwillard I removed `Argmax`\`s `R_op` and replace` Value Error` on ` aesara.gradient.grad_undefined`.

**Comment by aerubanov** [comment]
@brandonwillard, just soft reminder)

File: aesara/tensor/math.py [hunk 14]
```diff
@@ -622,11 +401,166 @@ def _c_all(self, node, name, inames, onames, sub):
         return decl, checks, alloc, loop, end
 
 
-class Max(NonZeroCAReduce):
-    nfunc_spec = ("max", 1, 1)
+class Max(COp):
```
Comment (brandonwillard) [inline aesara/tensor/math.py:404]: Why aren't we using the `NonZeroCAReduce` base class for this?

File: tests/tensor/test_math.py [hunk 15]
```diff
@@ -747,11 +747,12 @@ def test_isnan():
 
 class TestMaxAndArgmax:
     def setup_method(self):
-        MaxAndArgmax.debug = 0
+        Argmax.debug = 0
+        Max.debug = 0
```
Comment (brandonwillard) [inline tests/tensor/test_math.py:751]: What do these do?

**Review by brandonwillard — COMMENTED** [review]
Looks like one of the commit messages got scrambled.

File: aesara/tensor/math.py (hunk 14 above)
Comment (aerubanov) [reply aesara/tensor/math.py:404]: @brandonwillard  Thank you for review. To fix this https://github.com/aesara-devs/aesara/pull/874#issuecomment-1118590762 problem with `Max` I coped the code from `MaxAndArgmax` and made it  `COp`. Do you think that it can add some problems?

File: tests/tensor/test_math.py (hunk 15 above)
Comment (aerubanov) [reply tests/tensor/test_math.py:751]: Hm, it looks like a piece of old code.  I think I should to remove it.

File: aesara/tensor/math.py (hunk 14 above)
Comment (brandonwillard) [reply aesara/tensor/math.py:404]: It could interfere with existing code that assumes `Max` is a subclass of `NonZeroCAReduce` (e.g. rewrites, our JAX and/or Numba implementations, etc.)

Also, we generally want to make use of existing code.  In this case, subclassing `NonZeroCAReduce` could make adding a C implementation much easier that it otherwise would be.
